source/Plugins.cpp

```diff
@@ -0,0 +1,135 @@
+/* Plugins.cpp
+Copyright (c) 2022 by Sam Gleske (samrocketman on GitHub)
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Plugins.h"
+
+#include "DataFile.h"
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Files.h"
+
+#include <algorithm>
+#include <map>
+
+using namespace std;
+
+namespace {
+	// The first item in pair is plugin state on load, second item is user changing preference.
+	map<string, pair<bool, bool>> settings;
+
+	void LoadSettings(const string &path)
+	{
+		DataFile prefs(path);
+		for(const DataNode &node : prefs)
+		{
+			const string &key = node.Token(0);
+			if(key == "plugins")
+				for(const DataNode &child : node)
+					if(child.Size() == 2)
+						settings[child.Token(0)].first = child.Value(1);
+		}
+	}
+}
+
+
+
+void Plugins::Load()
+{
+	// Global plugin settings
+	LoadSettings(Files::Resources() + "plugins.txt");
+	// Local plugin settings
+	LoadSettings(Files::Config() + "plugins.txt");
+}
+
+
+
+void Plugins::Save()
+{
+	if(settings.empty())
+		return;
+	DataWriter out(Files::Config() + "plugins.txt");
+
+	out.Write("plugins");
+	out.BeginChild();
+	{
+		for(const auto &it : settings)
+			out.Write(it.first, it.second.second);
+	}
+	out.EndChild();
+}
+
+
+
+// Freeze the plugin set to determine if a setting has changed.
+void Plugins::Freeze()
+{
+	for(auto &it : settings)
+		it.second.second = it.second.first;
+}
+
+
+
+// Check if a plugin is known. It does not matter if it is enabled or disabled.
+bool Plugins::Has(const string &name)
+{
+	auto it = settings.find(name);
+	return it != settings.end();
+}
+
+
+
+// Determine if a plugin setting has changed since launching.
+bool Plugins::HasChanged(const string &name)
+{
+	const auto it = settings.find(name);
+	return (it == settings.end()) || it->second.first != it->second.second;
```

Comment:
Shouldn't this return false when the value is not in the map? (The plugin has definitely not changed state if we didn't know about it.)

```suggestion
	return it == settings.end() ? false :  it->second.first != it->second.second;
```

Reply:
The idea was, if the plugin isn't in the map, something has gone wrong and you should probably restart anyway.

Reply:
This is more of an edge case.  In reality it should be impossible to ever reach `settings.end()` so at best reaching that point should he some kind of error.  For now, the way the error is surfaced for this case is to warn the user to restart.

Reply:
HasChanged == true is intentional when `settings.end()` is reached.  I want to keep `settings.end()` protection to prevent crashes.  But in general it shouldn't ever return `true`.